<?xml version="1.0" encoding="UTF-8"?>
<mule xmlns="http://www.mulesoft.org/schema/mule/core"
      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
      xmlns:http="http://www.mulesoft.org/schema/mule/http"
      xmlns:test="http://www.mulesoft.org/schema/mule/test"
      xmlns:spring="http://www.springframework.org/schema/beans"
      xsi:schemaLocation="http://www.springframework.org/schema/beans http://www.springframework.org/schema/beans/spring-beans-current.xsd
       http://www.mulesoft.org/schema/mule/core http://www.mulesoft.org/schema/mule/core/current/mule.xsd
       http://www.mulesoft.org/schema/mule/test http://www.mulesoft.org/schema/mule/test/current/mule-test.xsd
       http://www.mulesoft.org/schema/mule/http http://www.mulesoft.org/schema/mule/http/current/mule-http.xsd">

    <spring:bean id="objectStore" class="org.mule.runtime.core.util.store.PartitionedPersistentObjectStore" />
    <spring:bean id="objectStore2" class="org.mule.runtime.core.util.store.PartitionedPersistentObjectStore" />

    <flow name="customProcessorInFlow">
        <custom-processor class="org.mule.test.integration.exceptions.ExceptionHandlingTestCase$ExceptionHandlerVerifierProcessor" />
    </flow>

    <flow name="asyncInFlow">
        <async>
            <custom-processor class="org.mule.test.integration.exceptions.ExceptionHandlingTestCase$ExceptionHandlerVerifierProcessor" />
            <test:component throwException="true"/>
        </async>
        <error-handler>
            <on-error-continue>
                <test:queue name="outFlow4"/>
            </on-error-continue>
        </error-handler>
    </flow>

    <flow name="untilSuccessfulInFlow">
        <until-successful objectStore-ref="objectStore" maxRetries="1" millisBetweenRetries="1000">
            <custom-processor class="org.mule.test.integration.exceptions.ExceptionHandlingTestCase$ExceptionHandlerVerifierProcessor" />
        </until-successful>
        <until-successful objectStore-ref="objectStore" maxRetries="1" millisBetweenRetries="1000">
            <test:component throwException="true"/>
        </until-successful>
        <error-handler>
            <on-error-continue>
                <test:queue name="outFlow5"/>
            </on-error-continue>
        </error-handler>
    </flow>

    <flow name="customProcessorInScope">
        <foreach>
            <custom-processor class="org.mule.test.integration.exceptions.ExceptionHandlingTestCase$ExceptionHandlerVerifierProcessor" />
        </foreach>
        <error-handler>
            <on-error-propagate/>
        </error-handler>
    </flow>

    <flow name="customProcessorInTransactionalScope">
        <transactional>
            <custom-processor class="org.mule.test.integration.exceptions.ExceptionHandlingTestCase$ExceptionHandlerVerifierProcessor" />
            <test:component throwException="true"/>
            <error-handler>
                <on-error-continue>
                    <test:assert expression="#[flowVars.expectedHandler == false]"/>
                    <test:queue name="outTransactional1"/>
                </on-error-continue>
            </error-handler>
        </transactional>
    </flow>

    <flow name="asyncInTransactionalScope">
        <transactional>
            <async>
                <custom-processor class="org.mule.test.integration.exceptions.ExceptionHandlingTestCase$ExceptionHandlerVerifierProcessor" />
                <test:component throwException="true"/>
            </async>
            <error-handler>
                <on-error-continue>
                    <test:queue name="outTransactional4"/>
                </on-error-continue>
            </error-handler>
        </transactional>
    </flow>

    <flow name="untilSuccessfulInTransactionalScope">
        <transactional>
            <until-successful objectStore-ref="objectStore" maxRetries="1" millisBetweenRetries="1000">
                <custom-processor class="org.mule.test.integration.exceptions.ExceptionHandlingTestCase$ExceptionHandlerVerifierProcessor" />
            </until-successful>
            <until-successful objectStore-ref="objectStore2" maxRetries="1" millisBetweenRetries="1000">
                <test:component throwException="true"/>
            </until-successful>
            <error-handler>
                <on-error-continue>
                    <test:queue name="outTransactional5"/>
                </on-error-continue>
            </error-handler>
        </transactional>
    </flow>

    <flow name="customProcessorInExceptionStrategy">
        <test:component throwException="true"/>
        <error-handler>
            <on-error-continue>
                <custom-processor class="org.mule.test.integration.exceptions.ExceptionHandlingTestCase$ExceptionHandlerVerifierProcessor" />
                <test:assert expression="#[flowVars.expectedHandler == false]"/>
                <test:queue name="outStrategy1"/>
            </on-error-continue>
        </error-handler>
    </flow>

    <flow name="asyncInExceptionStrategy">
        <test:component throwException="true"/>
        <error-handler>
            <on-error-continue>
                <custom-processor class="org.mule.test.integration.exceptions.ExceptionHandlingTestCase$ExecutionCountProcessor" />
                <async>
                    <custom-processor class="org.mule.test.integration.exceptions.ExceptionHandlingTestCase$ExceptionHandlerVerifierProcessor" />
                    <test:component throwException="true"/>
                </async>
            </on-error-continue>
        </error-handler>
    </flow>

    <flow name="untilSuccessfulInExceptionStrategy">
        <test:component throwException="true"/>
        <error-handler>
            <on-error-continue>
                <custom-processor class="org.mule.test.integration.exceptions.ExceptionHandlingTestCase$ExecutionCountProcessor" />
                <until-successful objectStore-ref="objectStore" maxRetries="1" millisBetweenRetries="1000">
                    <custom-processor class="org.mule.test.integration.exceptions.ExceptionHandlingTestCase$ExceptionHandlerVerifierProcessor" />
                </until-successful>
                <until-successful objectStore-ref="objectStore" maxRetries="1" millisBetweenRetries="1000">
                    <test:component throwException="true"/>
                </until-successful>
            </on-error-continue>
        </error-handler>
    </flow>

    <flow name="untilSuccessfulInExceptionStrategyRollback">
        <test:component throwException="true"/>
        <error-handler>
            <on-error-propagate>
                <custom-processor class="org.mule.test.integration.exceptions.ExceptionHandlingTestCase$ExecutionCountProcessor" />
                <until-successful objectStore-ref="objectStore" maxRetries="1" millisBetweenRetries="1000">
                    <custom-processor class="org.mule.test.integration.exceptions.ExceptionHandlingTestCase$ExceptionHandlerVerifierProcessor" />
                </until-successful>
                <until-successful objectStore-ref="objectStore" maxRetries="1" millisBetweenRetries="1000">
                    <test:component throwException="true"/>
                </until-successful>
            </on-error-propagate>
        </error-handler>
    </flow>

</mule>
